pkg/chaosdaemon/server.go

```diff
@@ -159,45 +154,70 @@ type RegisterGatherer interface {
 	prometheus.Gatherer
 }
 
-// StartServer starts chaos-daemon.
-func StartServer(conf *Config, reg RegisterGatherer) error {
-	g := &errgroup.Group{}
+// Server is the server for chaos daemon
+type Server struct {
+	errors       chan error
+	daemonServer *DaemonServer
 
-	httpBindAddr := conf.HttpAddr()
-	httpServer := newHTTPServerBuilder().Addr(httpBindAddr).Metrics(reg).Profiling(conf.Profiling).Build()
+	conf *Config
+	reg  RegisterGatherer
+}
+
+// BuildServer builds a chaos daemon server
+func BuildServer(conf *Config, reg RegisterGatherer) (*Server, error) {
+	daemonServer, err := newDaemonServer(conf.Runtime, reg)
+	if err != nil {
+		return nil, errors.Wrap(err, "create daemon server")
+	}
 
-	grpcBindAddr := conf.GrpcAddr()
+	return &Server{
+		errors:       make(chan error),
+		daemonServer: daemonServer,
+		conf:         conf,
+		reg:          reg,
+	}, nil
+}
+
+// Start starts chaos-daemon.
+func (s *Server) Start() error {
+	httpBindAddr := s.conf.HttpAddr()
+	httpServer := newHTTPServerBuilder().Addr(httpBindAddr).Metrics(s.reg).Profiling(s.conf.Profiling).Build()
+
+	grpcBindAddr := s.conf.GrpcAddr()
 	grpcListener, err := net.Listen("tcp", grpcBindAddr)
 	if err != nil {
-		log.Error(err, "failed to listen grpc address", "grpcBindAddr", grpcBindAddr)
-		return err
+		return errors.Wrapf(err, "listen grpc address %s", grpcBindAddr)
 	}
 
-	grpcServer, err := newGRPCServer(conf.Runtime, reg, conf.tlsConfig)
+	grpcServer, err := newGRPCServer(s.daemonServer, s.reg, s.conf.tlsConfig)
 	if err != nil {
-		log.Error(err, "failed to create grpc server")
-		return err
+		return errors.Wrap(err, "create grpc server")
 	}
 
-	g.Go(func() error {
+	go func() {
 		log.Info("Starting http endpoint", "address", httpBindAddr)
 		if err := httpServer.ListenAndServe(); err != nil {
-			log.Error(err, "failed to start http endpoint")
 			httpServer.Shutdown(context.Background())
-			return err
+			s.errors <- errors.Wrap(err, "start http endpoint")
 		}
-		return nil
-	})
+	}()
 
-	g.Go(func() error {
-		log.Info("Starting grpc endpoint", "address", grpcBindAddr, "runtime", conf.Runtime)
+	go func() {
+		log.Info("Starting grpc endpoint", "address", grpcBindAddr, "runtime", s.conf.Runtime)
 		if err := grpcServer.Serve(grpcListener); err != nil {
-			log.Error(err, "failed to start grpc endpoint")
 			grpcServer.Stop()
-			return err
+			s.errors <- errors.Wrap(err, "start grpc endpoint")
 		}
-		return nil
-	})
+	}()
+
+	return nil
+}
+
+// Errors return the error channel
+func (s *Server) Errors() <-chan error {
+	return s.errors
+}
 
-	return g.Wait()
+func (s *Server) Shutdown() {
+	s.daemonServer.backgroundProcessManager.Shutdown()
```

Comment:
we should also "Close" other resources like `grpc` and `http` server

Reply:
fine

Reply:
Fixed